src/ripple/ledger/impl/ApplyView.cpp

```diff
@@ -0,0 +1,293 @@
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+#include <ripple/ledger/ApplyView.h>
+#include <ripple/basics/contract.h>
+#include <ripple/protocol/Protocol.h>
+#include <cassert>
+
+namespace ripple {
+
+boost::optional<std::uint64_t>
+ApplyView::dirAdd (
+    Keylet const& directory,
+    uint256 const& key,
+    std::function<void(std::shared_ptr<SLE> const&)> const& describe,
+    std::function<void(STVector256&, uint256 const&)> const& add)
+{
+    auto root = peek(directory);
+
+    if (! root)
+    {
+        // No root, make it.
+        root = std::make_shared<SLE>(directory);
+        root->setFieldH256 (sfRootIndex, directory.key);
+        describe (root);
+
+        STVector256 v;
+        v.push_back (key);
+        root->setFieldV256 (sfIndexes, v);
+
+        insert (root);
+        return std::uint64_t{0};
+    }
+
+    std::uint64_t page = root->getFieldU64(sfIndexPrevious);
+
+    auto node = root;
+
+    if (page)
+    {
+        node = peek (keylet::page(directory, page));
+        if (!node)
+            LogicError ("Directory chain: root back-pointer broken.");
+    }
+
+    auto indexes = node->getFieldV256(sfIndexes);
+
+    // If there's space, we use it:
+    if (indexes.size () < dirNodeMaxEntries)
+    {
+        add (indexes, key);
+        node->setFieldV256 (sfIndexes, indexes);
+        update(node);
+        return page;
+    }
+
+    // Check whether we're out of pages.
+    if (++page >= dirNodeMaxPages)
+        return boost::none;
+
+    // We are about to create a new node; we'll link it to
+    // the chain first:
+    node->setFieldU64 (sfIndexNext, page);
+    update(node);
+
+    root->setFieldU64 (sfIndexPrevious, page);
+    update(root);
+
+    // Insert the new key:
+    indexes.clear();
+    indexes.push_back (key);
+
+    node = std::make_shared<SLE>(keylet::page(directory, page));
+    node->setFieldH256 (sfRootIndex, directory.key);
+    node->setFieldV256 (sfIndexes, indexes);
+
+    // Save some space by not specifying the value 0 since
+    // it's the default.
+    if (page != 1)
+        node->setFieldU64 (sfIndexPrevious, page - 1);
+    describe (node);
+    insert (node);
+
+    return page;
+}
+
+boost::optional<std::uint64_t>
+ApplyView::dirInsert (
+    Keylet const& directory,
+    uint256 const& key,
+    std::function<void (std::shared_ptr<SLE> const&)> const& describe)
+{
+    return dirAdd (directory, key, describe,
+        [](STVector256& v, uint256 const& k)
+        {
+            // We can't be sure if this page is already sorted because
+            // it may be a legacy page we haven't yet touched. Take
+            // the time to sort it.
+            std::sort (v.begin(), v.end());
+
+            auto pos = std::lower_bound(v.begin(), v.end(), k);
+
+            if (pos != v.end() && k == *pos)
+                LogicError ("dirInsert: double insertion");
+
+            v.insert (pos, k);
+        });
```

Comment:
One more quick note on this: If you want to keep the `std::function<void(STVector256&, uint256 const&)> const& add` parameter on `dirAdd`, consider making it a template instead of a `std::function`. Rational: using `std::function` incurs a virtual function call, and may require allocation (though since our lambdas don't have captures, unlikely in our cases). Since it's private, and only called from the `cpp`, we can easily declare it in the header and define it in the cpp.